fms/utils/generation.py

```diff
@@ -4,6 +4,88 @@
 import torch.nn.functional as F
 
 
+def __create_prefill_mask(
+    prompt_lengths: List[int], device: Union[str, torch.device]
+) -> torch.Tensor:
+    """Create a prefill mask based on prompt_lengths
+
+    Args:
+        prompt_lengths: List[int]
+            list of prompt lengths
+        device: Union[str, torch.device]
+            device to put mask on
+
+    Returns:
+    torch.Tensor
+        a causal mask
+    """
+    max_tokens = max(prompt_lengths)
+
+    is_pad_list = []
+    for seq_len in prompt_lengths:
+        pads = torch.zeros(max_tokens - seq_len, dtype=torch.bool, device=device)
+        non_pads = torch.ones(seq_len, dtype=torch.bool, device=device)
+        is_pad_list.append(torch.cat((pads, non_pads)))
+    is_pad = torch.stack(is_pad_list)
+    mask = is_pad.unsqueeze(-1) == is_pad.unsqueeze(-2)
+    return mask.tril(diagonal=0)
+
+
+def __create_decode_mask(
+    context_lengths: List[int], device: Union[str, torch.device]
+) -> torch.Tensor:
+    """create a decode mask
```

Comment:
same re what brian said, just say a 2..n decoding iteration mask (also know as a decoding step)

Reply:
added more documentation in latest commit